ingestion-beam/src/main/java/com/mozilla/telemetry/contextualservices/AggregateImpressions.java

```diff
@@ -0,0 +1,109 @@
+package com.mozilla.telemetry.contextualservices;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.Iterables;
+import com.mozilla.telemetry.ingestion.core.Constant.Attribute;
+import com.mozilla.telemetry.transforms.PubsubConstraints;
+import com.mozilla.telemetry.util.Json;
+import com.mozilla.telemetry.util.Time;
+import java.util.List;
+import java.util.Map;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.GroupByKey;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.GlobalWindows;
+import org.apache.beam.sdk.transforms.windowing.IntervalWindow;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.joda.time.Duration;
+
+/**
+ * Count the number of impressions aggregated by reporting URL.
+ */
+public class AggregateImpressions
+    extends PTransform<PCollection<PubsubMessage>, PCollection<PubsubMessage>> {
+
+  private final List<String> aggregationFields = ImmutableList.of(
+      ReportingUrlUtil.PARAM_COUNTRY_CODE, ReportingUrlUtil.PARAM_REGION_CODE,
+      ReportingUrlUtil.PARAM_FORM_FACTOR, ReportingUrlUtil.PARAM_OS_FAMILY,
+      ReportingUrlUtil.PARAM_ID);
+
+  private final ValueProvider<Integer> getAggregationWindowSize;
+
+  public AggregateImpressions(ValueProvider<Integer> getAggregationWindowSize) {
+    this.getAggregationWindowSize = getAggregationWindowSize;
+  }
+
+  public static AggregateImpressions of(ValueProvider<Integer> getAggregationWindowSize) {
+    return new AggregateImpressions(getAggregationWindowSize);
+  }
+
+  @Override
+  public PCollection<PubsubMessage> expand(PCollection<PubsubMessage> messages) {
+    return messages.apply(ParDo.of(new AddAggregationKey())) //
```

Comment:
I need to do some more reading through Beam docs, but this looks like a reasonable track. We should probably drop the message at this point and just have a KV with an empty value, so that we don't have to shuffle the value. There's a `Count.perKey()` transform we may want to use here.